### ISSUE TITLE
Debug decoding failing with UnmappableCharacterException

### ISSUE THREAD

**Issue by borgesgabriel** [iBotPeaches/Apktool#1038]
When decoding a `.apk` with the debug (`-d`) flag on, I encounter an UnmappableCharacterException. This does not happen when such flag is off.
The project has been obfuscated with Dexguard; many classes are named with single non-western Unicode characters.
The interesting bit of the output of `java -jar apktool.jar d app.apk -d` is:
```
I: Baksmaling classes.dex...
Exception in thread "main" brut.androlib.AndrolibException: brut.androlib.AndrolibException: java.nio.charset.UnmappableCharacterException: Input length = 1
        at brut.androlib.Androlib.decodeSourcesSmali(Androlib.java:96)
        at brut.androlib.ApkDecoder.decode(ApkDecoder.java:129)
        at brut.apktool.Main.cmdDecode(Main.java:165)
        at brut.apktool.Main.main(Main.java:81)
Caused by: brut.androlib.AndrolibException: java.nio.charset.UnmappableCharacterException: Input length = 1
        at brut.androlib.src.SmaliDecoder.decode(SmaliDecoder.java:111)
        at brut.androlib.src.SmaliDecoder.decode(SmaliDecoder.java:46)
        at brut.androlib.Androlib.decodeSourcesSmali(Androlib.java:94)
        ... 3 more
Caused by: java.nio.charset.UnmappableCharacterException: Input length = 1
        at java.nio.charset.CoderResult.throwException(Unknown Source)
        at sun.nio.cs.StreamDecoder.implRead(Unknown Source)
        at sun.nio.cs.StreamDecoder.read(Unknown Source)
        at java.io.InputStreamReader.read(Unknown Source)
        at java.io.BufferedReader.fill(Unknown Source)
        at java.io.BufferedReader.readLine(Unknown Source)
        at java.io.BufferedReader.readLine(Unknown Source)
        at brut.androlib.src.SmaliDecoder$SmaliFileVisitor.visitFile(SmaliDecoder.java:142)
        at brut.androlib.src.SmaliDecoder$SmaliFileVisitor.visitFile(SmaliDecoder.java:123)
        at java.nio.file.Files.walkFileTree(Unknown Source)
        at java.nio.file.Files.walkFileTree(Unknown Source)
        at brut.androlib.src.SmaliDecoder.decode(SmaliDecoder.java:108)
        ... 5 more
```
I can gladly link the `.apk` if needed.

**Comment by iBotPeaches** [comment]
I can't seem to duplicate this, so I would appreciate an APK. If its private you can email it to me (connor (dot) tumbleson (at) gmail (dot) com).

Or just upload it here / any file hosting site.

**Comment by borgesgabriel** [comment]
Just left you an email with the apk!
In other news, installing the apk back again after a disassemble and build with no modifications done to the code crashes my phone, when done via adb. I have confirmed that signature verification and the like are indeed disabled. Not sure if problems are related, though.

**Comment by iBotPeaches** [comment]
```
ibotpeaches@raganok:~/Downloads/Apktool/Bug1038$ apktool d -d xxx.apk 
I: Using Apktool 2.0.2-421999-SNAPSHOT on xxx.apk
I: Loading resource table...
I: Decoding AndroidManifest.xml with resources...
I: Loading resource table from file: /home/ibotpeaches/apktool/framework/1.apk
I: Regular manifest package...
I: Decoding file-resources...
I: Decoding values */* XMLs...
I: Baksmaling classes.dex...
I: Copying assets and libs...
I: Copying unknown files...
I: Copying original files...
ibotpeaches@raganok:~/Downloads/Apktool/Bug1038$ apktool b -d xxx
I: Using Apktool 2.0.2-421999-SNAPSHOT
I: Checking whether sources has changed...
I: Smaling smali folder into classes.dex...
I: Checking whether resources has changed...
I: Building resources...
I: Copying libs... (/lib)
I: Building apk file...
I: Copying unknown files/dir...
```

I didn't have any trouble here. (I haven't tested the crash yet). Can I ask what OS you are on?

If unix can I have output of

```
echo $LC_CTYPE
echo $LANG
```

**Comment by borgesgabriel** [comment]
I had tested it on both Linux and Windows, on a "en_GB" locale.
However, moving to Fedora on a "en_US" locale did the trick, and I was able to decode and build the apk with the `-d` flag on. Haven't yet been able to test whether the apk wasn't corrupted or anything because I'm still trying to bypass Dexguard's tamper detection mechanisms. That, however, is out of the scope of this issue.

It seems the problem was in fact with my system, hence I see no reason for this to remain open. Thanks for your help.
